### PR TITLE
Indexed Dynamic Priority Queue.

### PR DESCRIPTION
**Pull request by mayankshriv**
1. Implemented priority queues with indexing (O(1) search on integer
key). Also supports dynamic updates of values, ie value ordering
also updated (as per priority). This will be used in trimming of server
side (index level) group-by results.
2. Seperate implementations for int key with double or Object value, ie,
primitive data type based instead of generics, for performance.
3. Unit tests for the same.

### CODE REVIEW

File: pinot-core/src/main/java/com/linkedin/pinot/core/util/IntDoubleIndexedPriorityQueue.java [hunk 1]
```diff
@@ -0,0 +1,343 @@
...
+  Int2IntOpenHashMap _keyToIndexMap;
+  Int2IntOpenHashMap _indexToKeyMap;
+  private boolean _minHeap;
+
+  /**
+   * Constructor for the class.
+   *
+   * @param minHeap Min order, ie smallest element on top.
+   */
+  public IntDoubleIndexedPriorityQueue(boolean minHeap) {
```
Comment (kishoreg) [inline pinot-core/src/main/java/com/linkedin/pinot/core/util/IntDoubleIndexedPriorityQueue.java]: Don't you want to take maxSize/initialSize

File: pinot-core/src/main/java/com/linkedin/pinot/core/util/IntDoubleIndexedPriorityQueue.java [hunk 2]
```diff
@@ -0,0 +1,343 @@
...
+   *   <li> Throws exception if the specified key does not exist. </li>
+   *   <li> Runtime complexity of O(1). </li>
+   * </ul>
+   *
+   * @param key Key for which to return the value
+   * @return Value for the key
+   */
+  public IntDoublePair get(int key) {
+    if (!_keyToIndexMap.containsKey(key)) {
+      throw new RuntimeException("Key not found: " + key);
```
Comment (kishoreg) [inline pinot-core/src/main/java/com/linkedin/pinot/core/util/IntDoubleIndexedPriorityQueue.java]: why not return null

File: pinot-core/src/main/java/com/linkedin/pinot/core/util/IntDoubleIndexedPriorityQueue.java [hunk 3]
```diff
@@ -0,0 +1,343 @@
...
+   *   <li> key+value pair is removed from the priority queue. </li>
+   *   <li> Throws runtime exception if the priority queue is empty. </li>
+   *   <li> Runtime complexity of O(1). </li>
+   * </ul>
+   *
+   * @return Key+Value pair
+   */
+  public IntDoublePair poll() {
+    if (isEmpty()) {
+      throw new RuntimeException("Empty collection, nothing to remove");
```
Comment (kishoreg) [inline pinot-core/src/main/java/com/linkedin/pinot/core/util/IntDoubleIndexedPriorityQueue.java]: same here, why not return null

File: pinot-core/src/main/java/com/linkedin/pinot/core/util/IntDoubleIndexedPriorityQueue.java [hunk 4]
```diff
@@ -0,0 +1,343 @@
...
+   * <ul>
+   *   <li> if v1 < v2, -1 for max, +1 for min mode. </li>
+   *   <li> if v1 > v2, -1 for max, -1 for min mode. </li>
+   *   <li> if v1 = v2,  0 for max, 0 for min mode. </li>
+   * </ul>
+   * @param v1 Value to compare
+   * @param v2 Value to compare
+   * @return Result of comparison (as described above).
+   */
+  private int compare(double v1, double v2) {
```
Comment (kishoreg) [inline pinot-core/src/main/java/com/linkedin/pinot/core/util/IntDoubleIndexedPriorityQueue.java]: better to use Double.compare(d1,d2) right? there are some nuances in comparing  doubles

Comment (kishoreg) [inline pinot-core/src/main/java/com/linkedin/pinot/core/util/IntDoubleIndexedPriorityQueue.java]: I saw this code elsewhere

Comment (mayankshriv) [inline pinot-core/src/main/java/com/linkedin/pinot/core/util/IntDoubleIndexedPriorityQueue.java]: Addressed all comments. Extracted out common code between the two PQ implementations into an abstract base class. While a lot of the code still 'seems' similar but it cannot be shared as one is based on primitive data types (for performance), while the other one is generic implementation.

Comment (kishoreg) [inline pinot-core/src/main/java/com/linkedin/pinot/core/util/IntDoubleIndexedPriorityQueue.java]: tx. Agree, it's a trade-off between performance and code duplication

**Pull request merged**